### ISSUE TITLE
Undefined index: button_priority

### ISSUE THREAD

**Issue by zebrahosting** [soflyy/oxygen-bugs-and-features#974]
Query monitor shows these errors in the front and backend:
Undefined index: button_priority

Backend:
    wp-content/plugins/oxygen/component-framework/api/oxygen.element.class.php:37
    OxygenElement->__construct()
    wp-content/plugins/oxygen/component-framework/api/wrapper/OxyEl.php:90
    OxyEl->__construct()
    wp-content/plugins/oxygen-woocommerce/elements/product-builder.php:155
    OxyWooCommerce->load_files()
    wp-content/plugins/oxygen-woocommerce/OxyWooCommerce.php:11
    OxyWooCommerce->__construct()
    wp-content/plugins/oxygen-woocommerce/plugin.php:32
    oxygen_woocommerce_init()
    wp-includes/class-wp-hook.php:287
    do_action('plugins_loaded')
    wp-settings.php:409

Front end:
wp-content/plugins/oxygen/component-framework/api/oxygen.element.class.php:37

**Comment by chris-castillo-dev** [comment]
@zebrahosting can you check this site in GTMetrix and tell me if you are experiencing a longer than normal server wait time on the initial page request?

I saw PHP errors similar to what you reported and saw the initial server wait time go from 130ms to between 400ms - 900ms. Just want to know if we can identify a clear correlation between the PHP errors and load times.

**Comment by zebrahosting** [comment]
We run our own VPS with NGINX, redis object cache and nginx pagecache. PHP 7.3 (tried 7.4)8G
Ubuntu 18.04LTS GB mem and SSD harddisk raid. MariaDB

TTFB is areound 115ms total loading time varies because of a lot of external calls (Active Campaign, Facebook , Analytics etc.
Just don't like to see so many errors in my sites. It will always have an effect somewhere.

**Comment by alexanderkladov** [comment]
I have the same issue on a local MAMP site.

The debug log file is filled with this repeated message:

```
PHP Notice:  Undefined index: button_priority in /Applications/MAMP/htdocs/dev/site/wp-content/plugins/oxygen/component-framework/api/oxygen.element.class.php on line 37
```

**Comment by Oxymatic** [comment]
Fresh Site here - Activate Oxygen Woocom plugin and throws a similar error on the front and back ends. 
Notice: Undefined index: button_priority in /var/www/wp-content/plugins/oxygen/component-framework/api/oxygen.element.class.php on line 37

**Comment by inferno-adam** [comment]
I have this issue too but I also have similar errors for icon-padding, icon-background-color and icon-color. I've tried running it in different environments on my local development environment as well as my remote development server and it didn't have any effect on the issue. I'm going to open a ticket with the Oxygen team to see if they can assist me with this. I'll report back if they can help me resolve the issue.

**Comment by jshacklette63** [comment]
> 
> 
> I have this issue too but I also have similar errors for icon-padding, icon-background-color and icon-color. I've tried running it in different environments on my local development environment as well as my remote development server and it didn't have any effect on the issue. I'm going to open a ticket with the Oxygen team to see if they can assist me with this. I'll report back if they can help me resolve the issue.

Did you have any luck with Oxygen?  I am also seeing this issue, started last night out of the blue.

**Comment by Oxymatic** [comment]
Oxy support suggested this wordaround: 

> Please add the following two lines to your wp-config.php file, just below the WP_DEBUG line:
> define( 'WP_DEBUG_DISPLAY', false );
> @ini_set( 'display_errors', 0 );

However, it took some messing around with the hosting provider as well, but did stop the error messages and Oxy woo is now working. Pretty much.

They acknowledged that it's a bug, not a feature, so there's hope for the future.

**Comment by alexanderkladov** [comment]
Oxygen's notices are an absolute mess.. I can't believe how long it's taking to solve this.

@Oxymatic this will just hide the errors, but they will still continue to flood the log file. So that's not ideal.

One solution would be to disable errors altogether. But the best temporary (while we wait for Oxygen team to get their notices act together) DIY solution I've seen was proposed by @cvladan 
[here](https://github.com/soflyy/oxygen-bugs-and-features/issues/883#issuecomment-627662275) - a small code snippet that selectively bypasses all Oxygen-specific notices at the error handler level. I added it via Code Snippets plugin & it worked for 99% of these obnoxious notices. Highly recommend.

Here is the snippet, to save everyone the click:

```php
// Bypass Oxygen-related notices
set_error_handler( function( $errno, $errstr, $errfile ) {
  return ( strpos( $errfile, 'plugins/oxygen' ) !== false );
}, E_NOTICE );
```

**Comment by zebrahosting** [comment]
Thanks for the snippet. I have also some code running to disable the debug message popping up in a release version and messing up pages for Safari users.
Indeed some work to do on that front. No need for more features, lets keep it clean and error free!

**Comment by jshacklette63** [comment]
I tried both code snippets above and am still not able to Edit templates.  For me, this problem is impacting my ability to edit ALL templates.  When I click edit, it just returns a blank white page with a 500 http error.

I cannot make any changes to my templates at all currently.

**Comment by alexanderkladov** [comment]
@jshacklette63 Have you tried using Firefox? I've heard some people saying that it works much better/faster with Oxygen.. I haven't tried it myself though, so can't be 100% sure.

**Comment by jshacklette63** [comment]
@alexanderkladov  Yeah, that is primary browser.  I have also tried others with no success.  It is very frustrating.

I have disabled all plugins for exception of Oxygen, Oxygen for Woo and WooCommerce itself. I have tried uninstalling and reinstalling Oxygen.  Have also completely moved the site to a fresh location and still no success.

**Comment by alexanderkladov** [comment]
@jshacklette63 Have you tried disabling all error logging altogether (instead of just disabling the display of errors like @Oxymatic suggested [here](https://github.com/soflyy/oxygen-bugs-and-features/issues/974#issuecomment-629137645))?

You can disable it by replacing `define('WP_DEBUG', true);` (or even if it's already `false`) in your `wp-config.php` with:

```php
// Disable display of errors and warnings in PHP
@ini_set( 'error_reporting', 0 );
@ini_set( 'display_errors', 0 );

// Disable all WordPress logging
define( 'WP_DEBUG', false );
define( 'WP_DEBUG_LOG', false );
define( 'SCRIPT_DEBUG', false );
define( 'SAVEQUERIES', false );
```

**Comment by jshacklette63** [comment]
@alexanderkladov I have tried everything mentioned in this thread including what you just included and still no luck.  Looks like it is hitting a loop and crashing out due to memory limitations.  Below are the errors I am getting when turning debugging back on for a moment. 

Notice: Undefined index: button_priority in /home/ti2mwr1jxocp/public_html/owp/wp-content/plugins/oxygen/component-framework/api/oxygen.element.class.php on line 37

Fatal error: Allowed memory size of 268435456 bytes exhausted (tried to allocate 126976 bytes) in /home/ti2mwr1jxocp/public_html/owp/wp-includes/functions.php on line 4609

The first error repeats several times until the memory is exhausted.

I tried commenting out the code for the button_priority (line 37 of oxygen.element.class.php) and even tried feeding it some data to make it shut up but nothing...  lol   It just leads to a larger chain of issues or breaks the code all together.

**Comment by alexanderkladov** [comment]
@jshacklette63 have you tried [this](https://github.com/soflyy/oxygen-bugs-and-features/issues/883#issuecomment-619428877)?

**Comment by jshacklette63** [comment]
@alexanderkladov I had not seen that one yet...  I applied the update to line 37 and I am no longer seeing that error!!!!  That is great!!  Thanks for sharing this!

However, now I am just getting this:

Fatal error: Allowed memory size of 268435456 bytes exhausted (tried to allocate 126976 bytes) in /home/ti2mwr1jxocp/public_html/owp/wp-includes/functions.php on line 4609

Any suggestion for this one?

Side note, thank you so much for taking the time to help me with this!

**Comment by KittenCodes** [comment]
@jshacklette63 

> Fatal error: Allowed memory size of 268435456 bytes exhausted (tried to allocate 126976 bytes) in /home/ti2mwr1jxocp/public_html/owp/wp-includes/functions.php on line 4609

Your site is running out of memory - please contact your web host and ask them to increase it for you.

**Comment by jshacklette63** [comment]
@KittenCodes While I was waiting for hosting support I ran across increasing the memory limits in Wordpress with:  define( 'WP_MAX_MEMORY_LIMIT', '512M' );

That got me past the memory error but then just ran into more... Got to below errors now:

PHP Notice:  Undefined index: progress_bar_background_color in /home/ti2mwr1jxocp/public_html/owp/wp-content/plugins/oxygen/component-framework/components/classes/progress-bar.class.php on line 219
PHP Notice:  Undefined index: progress_bar_bar_color in /home/ti2mwr1jxocp/public_html/owp/wp-content/plugins/oxygen/component-framework/components/classes/progress-bar.class.php on line 235

Tried the code from @alexanderkladov above to disable the display of errors and still cannot access the Oxygen Editor.  I have no clue what else to do here.  It appears that something corrupted the data that Oxygen is looking for in order to display the Oxygen editor.  That is the only thing I can think of that would cause multiple undefined indexes and the chain of errors.

**Comment by jshacklette63** [comment]
@KittenCodes & @alexanderkladov I went in poked around a bit more.  I am now able to load other Oxygen templates in the builder.  It is now just the main template that I cannot load.  However I am getting much closer.  I am now able to see the wordpress edit page and click the "Edit with Oxygen" button.  The Oxygen editor now starts the loading animation but then I get a 500 http error due to the errors I called out above I assume.  This is now where the editors breaks and stops loading.  So Close to getting past this!!!

**Comment by jshacklette63** [comment]
Alright, so I was able to get past this last issue by setting the max memory to 999M

define( 'WP_MAX_MEMORY_LIMIT', '999M' );

No clue why it is so memory hungry all of sudden but this change is now allowing me to access the Oxygen editor for all pages.

Thanks for your help as I went through this process!

**Comment by LauGau** [comment]
I am still having the issue : 
`PHP Notice:  Undefined index: button_priority in /.../htdocs/wp-content/plugins/oxygen/component-framework/api/oxygen.element.class.php on line 37
`
Still no fix for this ?

**Comment by eizil** [comment]
as of now, I simply add own definition for the index.

you can try this if you want, no warranty, if you have any problem, feel free to reach out to me.

add this line before the //set param (line 31)


`        $other_options['button_priority'] = isset($other_options['button_priority']) ? $other_options['button_priority'] : 10;
`

this might not be the best solution, but I just don't want to simply hide the error, when I know the error is there. At least with this, I set the index for button_priority. 

this is just a quick solution, since it's been months since I contacted oxygen support and they simply don't bothered to fix this issue.

**Comment by Jean-Like-Denim** [comment]
There was an issue for me that extended beyond just not seeing the problem, I wasn't able to open the editor with this error. After switching off any potential plugin conflicts I updated to Beta 3.5. 
My colour picker wasn't working either and the beta version sorted that out. The 'Undefined index: button_priority' issue persisted with the update so I commented the line out and now its working fine for me. I can open and use the editor and the colour picker is working.
